requests-for-comments/2024-01-23-project-environments.md

```diff
@@ -0,0 +1,189 @@
+
+#### Why not?
+
+Copying would mean a need to maintain such underlying links (e.g. a cross-project ID, adjacent to the primary key ID) for all entities we'd like to work across projects, which would be dashboards, tiles, insights, actions and cohorts at the minimum. In one way, this could be nice if you'd like to develop, say, a dashboard in the test environment and only copy it to the production environment when final, while still being able to diverge between projects. The major UX tradeoff would be the act of manual syncing, making the experience cumbersome for users at scale. This would work, but it wouldn't be turnkey. And the whole tree of dependencies would be massive in terms of entities in Postgres. Support reports around inconsistencies with dependency syncing would be hard to avoid.
+
+## Dismissed solution B: Dashboards/insights/etc. sharing between projects
+
+What if instead of copying whole trees of entities we shared them directly? This could be even less work for the user – no need to sync manually – and for us – no need to copy dozens of entities to share a single dashboard between two projects. Also no extra work in terms of permissioning or ingestion.
+
+#### Why not?
+
+The UX would still be a little clunky, with a need to manually set up multiple full projects for each environment. It's the taxonomy that's the killer though – shared taxonomy seems to be a relatively common ask in itself, but it's also key for analysis in an implicit way. Group type indexes in particular MUST match exactly, or insights will show results for completely different data. Property types matter – numbers are treated differently from strings, and then there are also dates. A deeper level of integration is needed for this to work reliably.
+
+## Proposed solution: Environments
+
+We can make environments a first-class entity in PostHog – each project would be subdivided into one or more environments. The *data* would be isolated between projects, but the *taxonomy* and *analysis setup* would be shared (see "Summarizing needs" for what should be isolated and what should be shared exactly).
+
+To support the most common workflow of drafting dashboards/etc. in dev/staging and then "promoting" to production, entities such as dashboards would be environment-specific by default, but could be rolled out to the whole project with two clicks. When rolled out, the entity's dependencies would be made project-wide as well (e.g. for a cohort, all actions and other cohorts used in the definition). This would involve no copying though, so there would be no effort needed to sync things.
```

Comment:
One potential complication that we should figure out a plan for in advance:

Essentially this means ability to query across `data_team_id` s for a query, which can be a bit dicey. Ex: a project-wide cohort used in a environment specific flag -> means you get the flag from the environment specific data_team_id and then the cohort from the 'production' data_team_id, because I assume if it's project-wide it doesn't get a second `environment_id` ? Or are you planning that whenever a new environment is added, to add a link from project-wide Cohorts, say, to also have the environment_id of the new environment? (and thus all cohorts that depend on it as well)

Reply:
.. oh and then environment permissions mean restricting some of these flows :oof-2:

Reply:
There is definitely no querying across `data_id` (fka `data_team_id`_ intended). I'm not sure what you mean with "link from project-wide Cohorts, say, to also have the environment_id of the new environment". If you make a cohort project-wide, we make all its dependencies project-wide too and ensure this stays that way. But it's not aware of multiple environments in any situation.

Reply:
So, cohort A is project-wide, which means all dependent cohorts become project wide.

But, then flag A depends on cohort A, so cohort A doesn't have a dependency on flag, but the flag does. And the flag can still be environment specific, even if cohort A is project-wide right?

Reply:
Yep